python/zygote.py

```diff
@@ -74,166 +75,177 @@ def try_dumps(obj, sort_keys=False, allow_nan=False):
-                    and arg_names[0] == "element_html"
-                    and arg_names[1] == "element_index"
-                    and arg_names[2] == "data"
-                ):
-                    args.insert(1, None)
-
-                # call the desired function in the loaded module
-                val = method(*args)
-
-                if fcn == "file":
-                    # if val is None, replace it with empty string
-                    if val is None:
-                        val = ""
-                    # if val is a file-like object, read whatever is inside
-                    if isinstance(val, io.IOBase):
-                        val.seek(0)
-                        val = val.read()
-                    # if val is a string, treat it as utf-8
-                    if isinstance(val, str):
-                        val = bytes(val, "utf-8")
-                    # if this next call does not work, it will throw an error, because
-                    # the thing returned by file() does not have the correct format
-                    val = base64.b64encode(val).decode()
-
-                # Any function that is not 'file' or 'render' will modify 'data' and
-                # should not be returning anything (because 'data' is mutable).
-                if (fcn != "file") and (fcn != "render"):
-                    if val is None:
-                        json_outp = try_dumps(
-                            {"present": True, "val": args[-1]}, allow_nan=False
-                        )
-                    else:
-                        json_outp_passed = try_dumps(
-                            {"present": True, "val": args[-1]},
-                            sort_keys=True,
-                            allow_nan=False,
-                        )
-                        json_outp = try_dumps(
-                            {"present": True, "val": val},
-                            sort_keys=True,
-                            allow_nan=False,
-                        )
-                        if json_outp_passed != json_outp:
-                            sys.stderr.write(
-                                'WARNING: Passed and returned value of "data" differ in the function '
-                                + str(fcn)
-                                + "() in the file "
-                                + str(cwd)
-                                + "/"
-                                + str(file)
-                                + ".py.\n\n passed:\n  "
-                                + str(args[-1])
-                                + "\n\n returned:\n  "
-                                + str(val)
-                                + '\n\nThere is no need to be returning "data" at all (it is mutable, i.e., passed by reference). In future, this code will throw a fatal error. For now, the returned value of "data" was used and the passed value was discarded.'
-                            )
+            data += s.recv(256 * 1024)
```

Comment:
So not sure what the purpose of this is, but I think data is a bytes object, which is immutable. The += and slice operators below will always make copies, which could get really slow. Better to use a bytearray.

Reply:
Would that just look like this?

```python
data = bytearray()
bytearray.extend(new_bytes)
```

Reply:
Yep, more or less a drop-in replacement with just a couple of different methods.

Reply:
Sweet - done in 1e196b5!